Jiade Song

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression

from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import MinMaxScaler
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
Auto = pd.read_csv('Auto.csv')
Auto[:5]

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0         130    3504          12.0    70   
1  15.0          8         350.0         165    3693          11.5    70   
2  18.0          8         318.0         150    3436          11.0    70   
3  16.0          8         304.0         150    3433          12.0    70   
4  17.0          8         302.0         140    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [4]:
Auto['mpg01']=np.nan
Auto

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0    18.0          8         307.0         130    3504          12.0    70   
1    15.0          8         350.0         165    3693          11.5    70   
2    18.0          8         318.0         150    3436          11.0    70   
3    16.0          8         304.0         150    3433          12.0    70   
4    17.0          8         302.0         140    3449          10.5    70   
..    ...        ...           ...         ...     ...           ...   ...   
387  27.0          4         140.0          86    2790          15.6    82   
388  44.0          4          97.0          52    2130          24.6    82   
389  32.0          4         135.0          84    2295          11.6    82   
390  28.0          4         120.0          79    2625          18.6    82   
391  31.0          4         119.0          82    2720          19.4    82   

     origin                       name  mpg01  
0         1  chevrolet chevelle malibu    NaN  
1         1          buick skylark 320    NaN  
2         1         plymouth satellite    NaN  
3         1              amc rebel sst    NaN  
4         1                ford torino    NaN  
..      ...                        ...    ...  
387       1            ford mustang gl    NaN  
388       2                  vw pickup    NaN  
389       1              dodge rampage    NaN  
390       1                ford ranger    NaN  
391       1                 chevy s-10    NaN  

[392 rows x 10 columns]

In [5]:
mpgmean = np.median(Auto.mpg)
mpgmean

22.75

In [6]:
length = len(Auto.mpg)
length

392

In [7]:
Auto.iloc[0,0]

18.0

In [8]:
for k in range(0,length):
    if Auto.mpg[k] > mpgmean:
        Auto.mpg01[k] = 1
    else :
        Auto.mpg01[k] = 0

In [9]:
Auto

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0    18.0          8         307.0         130    3504          12.0    70   
1    15.0          8         350.0         165    3693          11.5    70   
2    18.0          8         318.0         150    3436          11.0    70   
3    16.0          8         304.0         150    3433          12.0    70   
4    17.0          8         302.0         140    3449          10.5    70   
..    ...        ...           ...         ...     ...           ...   ...   
387  27.0          4         140.0          86    2790          15.6    82   
388  44.0          4          97.0          52    2130          24.6    82   
389  32.0          4         135.0          84    2295          11.6    82   
390  28.0          4         120.0          79    2625          18.6    82   
391  31.0          4         119.0          82    2720          19.4    82   

     origin                       name  mpg01  
0         1  chevrolet chevelle malibu    0.0  
1         1          buick skylark 320    0.0  
2         1         plymouth satellite    0.0  
3         1              amc rebel sst    0.0  
4         1                ford torino    0.0  
..      ...                        ...    ...  
387       1            ford mustang gl    1.0  
388       2                  vw pickup    1.0  
389       1              dodge rampage    1.0  
390       1                ford ranger    1.0  
391       1                 chevy s-10    1.0  

[392 rows x 10 columns]

predictor : cylinders, displacement, and weight

predicting : mpg01

In [10]:
Xcy = pd.DataFrame(Auto.cylinders)
Xdis = pd.DataFrame(Auto.displacement)
Xwei = pd.DataFrame(Auto.weight)
y = pd.DataFrame(Auto.mpg01)

In [11]:
model_cy = LogisticRegression(solver = 'lbfgs')
model_cy.fit(Xcy,y)

LogisticRegression()

In [12]:
model_cy.coef_

array([[-1.59892117]])

In [13]:
model_cy.intercept_

array([8.25279933])

In [14]:
model_dis = LogisticRegression(solver = 'lbfgs')
model_dis.fit(Xdis,y)

LogisticRegression()

In [15]:
model_dis.coef_

array([[-0.03272293]])

In [16]:
model_dis.intercept_

array([5.58577397])

In [17]:
model_wei = LogisticRegression(solver = 'lbfgs')
model_wei.fit(Xwei,y)

LogisticRegression()

In [18]:
model_wei.coef_

array([[-0.00411935]])

In [19]:
model_wei.intercept_

array([11.76328646])

\begin{equation}
log{(\frac{p(Xcy)}{1-p(Xcy)})}=8.253+(-1.599)Xcy
\end{equation}

\begin{equation}
log{(\frac{p(Xdis)}{1-p(Xdis)})}=5.586+(-0.033)Xdis
\end{equation}

\begin{equation}
log{(\frac{p(Xwei)}{1-p(Xwei)})}=11.763+(-0.0041)Xdis
\end{equation}

b)

In [20]:
Xcy_train,Xcy_test,y_train,y_test = \
train_test_split(Xcy,y,stratify=y,random_state=1)

Xdis_train,Xdis_test,y_train,y_test = \
train_test_split(Xdis,y,stratify=y,random_state=1)

Xwei_train,Xwei_test,y_train,y_test = \
train_test_split(Xwei,y,stratify=y,random_state=1)

In [21]:
m = len(y_test)
n = len(y)
m/n

0.25

In [22]:
model_cy1 = LogisticRegression(solver = 'lbfgs',max_iter=10000).fit(Xcy_train,y_train)
ycyhat = model_cy1.predict(Xcy_test)
model_cy1.score(Xcy_test,y_test)

0.8979591836734694

In [23]:
model_dis1 = LogisticRegression(solver = 'lbfgs',max_iter=10000).fit(Xdis_train,y_train)
ydishat = model_dis1.predict(Xdis_test)
model_dis1.score(Xdis_test,y_test)

0.8877551020408163

In [24]:
model_wei1 = LogisticRegression(solver = 'lbfgs',max_iter=10000).fit(Xwei_train,y_train)
yweihat = model_wei1.predict(Xwei_test)
model_wei1.score(Xwei_test,y_test)

0.9081632653061225

Test accuracy rates are 0.898 for culinders, 0.888 for displacements and 0.908 for weight. 

c).

In [25]:
data = Auto[['cylinders','displacement','weight']]
Xdata= pd.DataFrame(data)
Xdata

cylinders  displacement  weight
0            8         307.0    3504
1            8         350.0    3693
2            8         318.0    3436
3            8         304.0    3433
4            8         302.0    3449
..         ...           ...     ...
387          4         140.0    2790
388          4          97.0    2130
389          4         135.0    2295
390          4         120.0    2625
391          4         119.0    2720

[392 rows x 3 columns]

In [26]:
kfold = StratifiedKFold(n_splits = 10,shuffle = True,random_state=1)

In [27]:
scaler = MinMaxScaler()
model1 = LogisticRegression(solver = 'lbfgs')
pipe1 = Pipeline([('transformer1', scaler), ('estimator1', model1)])
scores1 = cross_val_score(pipe1,Xdata,y,cv=kfold)
scores1

array([0.95      , 0.9       , 0.92307692, 0.94871795, 0.92307692,
       0.87179487, 0.84615385, 0.8974359 , 0.92307692, 0.79487179])

In [28]:
scores1.mean()

0.8978205128205129

Test accuracy rate is 0.898. 

d).

In [29]:
Xdata_train,Xdata_test,y_train,y_test = \
train_test_split(Xdata,y,stratify=y,random_state=1)

In [30]:
scaler = MinMaxScaler()
model2 = KNeighborsClassifier(n_neighbors=3)
pipe2 = Pipeline([('transformer2', scaler), ('estimator2', model2)])
scores2 = cross_val_score(pipe2,Xdata_train,y_train)
print('Error Rate: '+str(1 - scores2.mean()))

Error Rate: 0.10204558737580349


e). 

In [31]:
acc_rate = []
nn = range(1,25)

In [32]:
kfold1 = StratifiedKFold(n_splits = 10,shuffle = True,random_state=1)

In [33]:
for k in nn:
    scaler = MinMaxScaler()
    model3 = KNeighborsClassifier(n_neighbors=k)
    pipe3 = Pipeline([('transformer3', scaler), ('estimator3', model3)])
    scores3 = cross_val_score(pipe3,Xdata,y,cv=kfold)
    acc_rate.append(scores3.mean())

In [34]:
print('Test error rate with the Best K: '+str(1 - max(acc_rate)))

Test error rate with the Best K: 0.08935897435897433
